kentik_api_library/kentik_api/analytics/flatness.py

```diff
@@ -0,0 +1,280 @@
+import io
+import json
+import logging
+import sys
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import pandas as pd  # type: ignore
+
+from kentik_api.utils import DeviceCache  # type: ignore
+
+from .data_frame_cache import DFCache  # type: ignore
+
+log = logging.getLogger("flatness_analysis")
+
+
+@dataclass
+class Interval:
+    start: datetime
+    end: datetime
+
+
+class FlatnessResults:
+    """
+    Class storing results of traffic flatness analysis on set of links
+    The events attribute contains dictionary keyed by link name containing list intervals in which traffic was
+    deemed to be 'flat'. Each interval is instance of 'Interval' named tuple.
+    """
+
+    def __init__(self, links: List[str]):
+        self.events: Dict[str, List[Interval]] = {link: [] for link in links}
+
+    def __getitem__(self, link):
+        return self.events.get(link)
+
+    @property
+    def stats(self) -> Dict[str, int]:
+        return dict(
+            total_events=sum(len(i) for i in self.events.values()),
+            max_per_link=max(len(i) for i in self.events.values()),
+        )
+
+    def last(self, link: str) -> Optional[Interval]:
+        if link in self.events and self.events[link]:
+            return self.events[link][-1]
+        else:
+            return None
+
+    def set_last(self, link: str, interval: Interval) -> None:
+        self.events[link][-1] = interval
```

Comment:
Should we protect against index out of bounds here?

Reply:
What would we do if it is out of bounds? Throw and exception? The code already does that :-). Since this can happen only due to a bug in the code, I think throwing `IndexError` (or `KeyError` in case the `link` key is not in the dictionary) is adequate for troubleshooting. The stack trace (hopefully) provides enough clue t or the programmer will have to use debugger to find the root cause. I don't think that an application specific exception would make troubleshooting any easier.

Reply:
Don't know the exact usage context of the function, but what comes to my mind if no last element:
- add new element
- log and ignore the call
Asking to make sure :)

Reply:
The situation simply won't happen with the code using this class today, because it guards against the condition. Any other application using the code will have to do the same. Moreover, the class is very specific to the needs of the "flatness analysis", so it is unlikely that it will be used in a different context. Regardless, the situation arises only due to a coding problem. It is better to let the application crash and provide useful information to the developer for debugging (i.e. stack trace). The invoking code simply has to handle this.  It would be a disservice to let execution continue if the problem occurs.